storage/innobase/include/fut0lst.h

```diff
@@ -117,11 +117,8 @@ inline uint32_t flst_get_len(const flst_base_node_t *base)
 /** @return a file address */
 inline fil_addr_t flst_read_addr(const byte *faddr)
 {
-  fil_addr_t addr= { mach_read_from_4(faddr + FIL_ADDR_PAGE),
-		     mach_read_from_2(faddr + FIL_ADDR_BYTE) };
-  ut_a(addr.page == FIL_NULL || addr.boffset >= FIL_PAGE_DATA);
```

Comment:
I have a suggestion here for your consideration. There are quite a bit of different code flows that reach this function. Since in Innodb the release asserts are the protection against further corruption, we should cautious while removing it. While the code path that are changed in this patch are very nicely done and covers the assert cases, there are other paths that are not.

One example is 
`trx_undo_seg_create -> flst_add_last -> flst_insert_after -> flst_get_next_addr -> flst_read_addr`

Instead of trying to solve all the cases together, can we take an incremental and safer approach ?

We have these four inline callers for the function.
`flst_get_first, flst_get_last, flst_get_next_addr, flst_get_prev_addr`

1. We create a new variant of these functions that don't assert.
   e.g. flst_get_first_no_assert or whatever name is appropriate. Perhaps, we could consider using template.

2. Whenever we handle the error cases 100%, we switch to using the new variant. e.g. for all cases handled in this patch.

In long run when there are no callers to the old functions, we eventually remove them.

Reply:
I would definitely not consider that for `flst_remove()` because the crash occurred there (while attempting to free a BLOB during the purge of committed transaction history). The general idea is that to deal with any corruption that was found while freeing something, we will ‘leak’ the data and not try to make the space reusable for future allocations.

As the commit message says, the operations that insert items into doubly linked lists do validate the location of the list node as well as the item that is being inserted. What will no longer be validated in these write paths is the existing list nodes that may be copied around. That is, if `flst_insert_after()` would copy its old `flst_get_next_addr()` to the list node that is being added. These would be validated on the read paths.

I realize that `flst_add_last()` and `flst_add_first()` are missing some opportunity to validate the data pointed to by `boffset`. I will revise them so that they will return `DB_CORRUPTION` if something is wrong.

Reply:
While revising this, I found out that `rseg->space->free_limit` would typically be incorrect when `trx_undo_mem_create_at_db_start()` is being invoked. Accessing the first page of `rseg->space` at the start of the function would prevent this check from tripping:
```c++
	fil_addr_t	last_addr = flst_get_last(
		TRX_UNDO_SEG_HDR + TRX_UNDO_PAGE_LIST + block->page.frame);

	if (last_addr.page >= rseg->space->free_limit
	    || last_addr.boffset < TRX_UNDO_PAGE_HDR + TRX_UNDO_PAGE_NODE
	    || last_addr.boffset >= srv_page_size
	    - TRX_UNDO_LOG_OLD_HDR_SIZE) {
	corrupted_undo:
		ut_free(undo);
		goto corrupted;
	}
```

Reply:
ffc0009aa6a9946e59bbcf70b2cd03a2a10cc424 addresses this by validating each page number against `fil_space_t::free_limit`.

Reply:
ffc0009aa6a9946e59bbcf70b2cd03a2a10cc424 adds the necessary access to the tablespace header page. It also revises the remaining entry points `flst_remove()`, `flst_add_first()`, and `flst_add_last()` so that `space->free_limit` will be passed as a parameter.